base/abstractarray.jl

```diff
@@ -2057,10 +2057,17 @@ end
 ## N argument
 
 @inline ith_all(i, ::Tuple{}) = ()
-@inline ith_all(i, as) = (as[1][i], ith_all(i, tail(as))...)
+function ith_all(i, as)
+    @_propagate_inbounds_meta
+    return (as[1][i], ith_all(i, tail(as))...)
+end
 
 function map_n!(f::F, dest::AbstractArray, As) where F
-    for i = LinearIndices(As[1])
+    #FIXME: adding `@_propagate_inbounds_meta` will destroy the performance of `map!`
+    #@_propagate_inbounds_meta
+    idxs1 = LinearIndices(As[1])
+    @boundscheck LinearIndices(dest) == idxs1 && all(x -> x == idxs1, map(LinearIndices, As))
```

Comment:
Could be written 
```suggestion
    @boundscheck LinearIndices(dest) == idxs1 && all(x -> LinearIndices(x) == idxs1, tail(As))
```

Reply:
```julia
julia> @btime foo!($a, $b, $c, $d, $e); # with tail
  371.754 ns (2 allocations: 80 bytes)

julia> @btime foo!($a, $b, $c, $d, $e); # without tail
  366.971 ns (1 allocation: 48 bytes)
```
`tail` introduces a little more allocation, I will use
```julia
all(x -> LinearIndices(x) == idxs1, As)
```
.